sympy/solvers/solveset.py

```diff
@@ -715,117 +525,120 @@ def _solve_abs(f, symbol):
+    elif f.is_Mul and all(_is_finite_with_finite_vars(m, domain)
+            for m in f.args):
+        # if f(x) and g(x) are both finite we can say that the solution of
+        # f(x)*g(x) == 0 is same as Union(f(x) == 0, g(x) == 0) is not true in
+        # general. g(x) can grow to infinitely large for the values where
+        # f(x) == 0. To be sure that we are not silently allowing any
+        # wrong solutions we are using this technique only if both f and g are
+        # finite for a finite input.
+        result = Union(*[solver(m, symbol) for m in f.args])
+    elif _is_function_class_equation(TrigonometricFunction, f, symbol) or \
+            _is_function_class_equation(HyperbolicFunction, f, symbol):
+        result = _solve_real_trig(f, symbol)
+    elif f.is_Piecewise:
+        dom = domain
         result = EmptySet()
-    elif f.is_Mul and all([_is_finite_with_finite_vars(m) for m in f.args]):
-        result = Union(*[solveset_complex(m, symbol) for m in f.args])
+        expr_set_pairs = f.as_expr_set_pairs()
+        for (expr, in_set) in expr_set_pairs:
+            if in_set.is_Relational:
+                in_set = in_set.as_set()
+            if in_set.is_Interval:
+                dom -= in_set
+            solns = solver(expr, symbol, in_set)
+            result += solns
     else:
-        lhs, rhs_s = invert_complex(f, 0, symbol)
+        lhs, rhs_s = inverter(f, 0, symbol)
         if lhs == symbol:
+            # do some very minimal simplification since
+            # repeated inversion may have left the result
+            # in a state that other solvers (e.g. poly)
+            # would have simplified; this is done here
+            # rather than in the inverter since here it
+            # is only done once whereas there it would
+            # be repeated for each step of the inversion
+            if isinstance(rhs_s, FiniteSet):
+                rhs_s = FiniteSet(*[Mul(*
+                    signsimp(i).as_content_primitive())
+                    for i in rhs_s])
             result = rhs_s
         elif isinstance(rhs_s, FiniteSet):
-            equations = [lhs - rhs for rhs in rhs_s]
-            result = EmptySet()
-            for equation in equations:
+            for equation in [lhs - rhs for rhs in rhs_s]:
                 if equation == f:
                     if any(_has_rational_power(g, symbol)[0]
                            for g in equation.args) or _has_rational_power(
                            equation, symbol)[0]:
                         result += _solve_radical(equation,
                                                  symbol,
-                                                 solveset_complex)
+                                                 solver)
+                    elif equation.has(Abs):
+                        result += _solve_abs(f, symbol, domain)
                     else:
                         result += _solve_as_rational(equation, symbol,
```

Comment:
The argument for `_solve_as_poly` also goes for `_solve_as_rational`, `domain` is enough to unambiguously specify the solvers.